test/e2e/virtual/apiexport/authorizer_test.go

```diff
@@ -345,19 +431,86 @@ func TestAPIExportAuthorizers(t *testing.T) {
 
 	t.Logf("verify that user-2 cannot lists CRD shadowed sherriffs resources in the tenant workspace %q via the virtual apiexport apiserver", tenantShadowCRDPath)
-	_, err = user2DynamicVWClient.Cluster(tenantShadowCRDClusterName.Path()).Resource(schema.GroupVersionResource{Version: "v1alpha1", Resource: "cowboys", Group: "wildwest.dev"}).List(ctx, metav1.ListOptions{})
+	_, err = user2DynamicVWClient.Cluster(logicalcluster.Name(tenantShadowCRDWorkspace.Spec.Cluster).Path()).Resource(schema.GroupVersionResource{Version: "v1alpha1", Resource: "cowboys", Group: "wildwest.dev"}).List(ctx, metav1.ListOptions{})
 	require.Error(t, err, "expected error, got none")
 	require.True(t, errors.IsNotFound(err))
 }
 
+var scheme *runtime.Scheme
+
+func init() {
+	scheme = runtime.NewScheme()
+	_ = apisv1alpha1.AddToScheme(scheme)
+	_ = rbacv1.AddToScheme(scheme)
+	_ = apiextensionsv1.AddToScheme(scheme)
+}
+
+func apply(t *testing.T, ctx context.Context, workspace logicalcluster.Path, cfg *rest.Config, manifests ...any) error {
+	t.Helper()
+	discoveryClient, err := kcpkubernetesclientset.NewForConfig(cfg)
+	require.NoError(t, err)
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(discoveryClient.Cluster(workspace).Discovery()))
+
+	for _, manifest := range manifests {
+		obj, gvk := func() (*unstructured.Unstructured, *schema.GroupVersionKind) {
+			switch value := manifest.(type) {
+			case string:
+				result, gvk, err := extensionsapiserver.Codecs.UniversalDeserializer().Decode([]byte(value), nil, &unstructured.Unstructured{})
+				require.NoError(t, err)
+
+				obj, ok := result.(*unstructured.Unstructured)
+				if !ok {
+					t.Fatalf("decoded into incorrect type, got %T, wanted %T", obj, &unstructured.Unstructured{})
+				}
+				return obj, gvk
+			case runtime.Object:
+				ro := manifest.(runtime.Object)
+				gvks, _, err := scheme.ObjectKinds(ro)
+				require.NoError(t, err)
+				ro.GetObjectKind().SetGroupVersionKind(gvks[0])
+				o, err := runtime.DefaultUnstructuredConverter.ToUnstructured(ro)
+				require.NoError(t, err)
+				return &unstructured.Unstructured{Object: o}, &gvks[0]
+			default:
+				panic("unsupported type")
+			}
+		}()
+
+		mapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+		require.NoError(t, err)
+
+		dynamicClient, err := kcpdynamic.NewForConfig(cfg)
+		require.NoError(t, err)
+		var dynamicResource dynamic.ResourceInterface
+		if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
+			t.Logf(`applying %q workspace %q namespace %q name %q`, gvk.String(), workspace.String(), obj.GetNamespace(), obj.GetName())
+			dynamicResource = dynamicClient.Cluster(workspace).Resource(mapping.Resource).Namespace(obj.GetNamespace())
+		} else {
+			t.Logf(`applying %q workspace %q name %q`, gvk.String(), workspace.String(), obj.GetName())
+			dynamicResource = dynamicClient.Cluster(workspace).Resource(mapping.Resource)
+		}
+
+		bytes, err := json.Marshal(obj)
```

Comment:
nit: use the scheme

Reply:
wdym, do you have a reference?

Reply:
You want a structured encoder from a scheme that has the type registered - I don't have a good link but would just auto-complete my way there.